File: data_converter.py
# data_converter.py
import re
import math
from collections import deque

# Unit definitions: (name, base_bytes, aliases)
UNITS = {
    'b':   ('bit', 1/8),
    'B':   ('byte', 1),
    'KB':  ('kilobyte', 1000),
    'KiB': ('kibibyte', 1024),
    'MB':  ('megabyte', 1000**2),
    'MiB': ('mebibyte', 1024**2),
    'GB':  ('gigabyte', 1000**3),
    'GiB': ('gibibyte', 1024**3),
    'TB':  ('terabyte', 1000**4),
    'TiB': ('tebibyte', 1024**4),
    'PB':  ('petabyte', 1000**5),
    'PiB': ('pebibyte', 1024**5),
}
UNIT_ALIASES = {u.lower(): u for u in UNITS}
for u in UNITS:
    UNIT_ALIASES[u.lower()] = u
    # add common aliases
    if u.endswith('B') and not u.startswith('b'):
        UNIT_ALIASES[u[:-1].lower()] = u  # e.g., 'kb' -> 'KB'

class DataConverter:
    def __init__(self, precision=2):
        self.precision = precision
        self.history = deque(maxlen=20)

    def parse_value_unit(self, s):
        """Parse string like '1.5 MB' or '2GiB' into (value, unit)."""
        s = s.strip()
        # match number and unit
        m = re.match(r'^([\d.]+)\s*([a-zA-Z]+)$', s)
        if not m:
            # try just number -> assume bytes
            try:
                val = float(s)
                return val, 'B'
            except:
                raise ValueError("Invalid format. Use 'value unit' (e.g., 1.5 MB).")
        val = float(m.group(1))
        unit_str = m.group(2).strip()
        unit = UNIT_ALIASES.get(unit_str.lower())
        if not unit:
            raise ValueError(f"Unknown unit: {unit_str}")
        return val, unit

    def to_bytes(self, value, unit):
        return value * UNITS[unit][1]

    def from_bytes(self, bytes_val, unit):
        return bytes_val / UNITS[unit][1]

    def convert(self, value, from_unit, to_unit):
        bytes_val = self.to_bytes(value, from_unit)
        result = self.from_bytes(bytes_val, to_unit)
        self.history.append((value, from_unit, result, to_unit))
        return result

    def auto_convert(self, value, unit, target_unit=None):
        """Convert, if target_unit is None, choose best human-readable unit."""
        bytes_val = self.to_bytes(value, unit)
        if target_unit:
            return self.from_bytes(bytes_val, target_unit), target_unit
        else:
            # find best unit: choose the one with value between 1 and 1000 (or 1024)
            best_unit = 'B'
            best_val = bytes_val
            for u, (name, factor) in UNITS.items():
                v = bytes_val / factor
                if 1 <= v < 1000:
                    best_unit = u
                    best_val = v
                    break
            # if all <1, pick smallest unit
            if best_unit == 'B' and bytes_val < 1:
                best_unit = 'b'
                best_val = bytes_val / UNITS['b'][1]
            return best_val, best_unit

    def format_result(self, val, unit, precision=None):
        if precision is None:
            precision = self.precision
        return f"{val:.{precision}f} {unit}"

    def add_history_line(self, line):
        self.history.append(line)

    def show_history(self):
        if not self.history:
            print("No conversions yet.")
        else:
            print("\n--- Conversion History (last 20) ---")
            for i, (v, fu, r, tu) in enumerate(self.history, 1):
                print(f"{i}. {v} {fu} -> {self.format_result(r, tu)}")

def main():
    converter = DataConverter(precision=2)
    print("=== Data Unit Converter ===")
    while True:
        print("\n1. Convert single value")
        print("2. Batch convert from file")
        print("3. Show conversion history")
        print("4. Set precision (current: {})".format(converter.precision))
        print("5. Help / unit info")
        print("6. Exit")
        choice = input("Choose: ").strip()
        if choice == '1':
            try:
                inp = input("Enter value and unit (e.g., 1024 MB): ")
                val, from_unit = converter.parse_value_unit(inp)
                target = input("Target unit (leave blank for auto): ").strip()
                if target:
                    target_unit = UNIT_ALIASES.get(target.lower())
                    if not target_unit:
                        print("Unknown target unit.")
                        continue
                    result = converter.convert(val, from_unit, target_unit)
                    print(f"\nResult: {converter.format_result(val, from_unit)} = {converter.format_result(result, target_unit)}")
                else:
                    result, best_unit = converter.auto_convert(val, from_unit)
                    print(f"\nResult: {converter.format_result(val, from_unit)} = {converter.format_result(result, best_unit)}")
            except Exception as e:
                print(f"Error: {e}")
        elif choice == '2':
            fname = input("Enter batch file path: ").strip()
            try:
                with open(fname, 'r') as f:
                    lines = f.readlines()
                target = input("Target unit for all conversions (leave blank for auto): ").strip()
                if target:
                    target_unit = UNIT_ALIASES.get(target.lower())
                    if not target_unit:
                        print("Unknown target unit.")
                        continue
                print("\nBatch results:")
                for line in lines:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        val, from_unit = converter.parse_value_unit(line)
                        if target:
                            result = converter.convert(val, from_unit, target_unit)
                            print(f"{converter.format_result(val, from_unit)} -> {converter.format_result(result, target_unit)}")
                        else:
                            result, best_unit = converter.auto_convert(val, from_unit)
                            print(f"{converter.format_result(val, from_unit)} -> {converter.format_result(result, best_unit)}")
                    except Exception as e:
                        print(f"Skipping '{line}': {e}")
            except FileNotFoundError:
                print("File not found.")
        elif choice == '3':
            converter.show_history()
        elif choice == '4':
            try:
                prec = int(input("Enter number of decimal places: "))
                if prec < 0:
                    raise ValueError
                converter.precision = prec
                print("Precision updated.")
            except:
                print("Invalid precision.")
        elif choice == '5':
            print("\nSupported units:")
            for u, (name, factor) in sorted(UNITS.items()):
                print(f"  {u:4} = {name} (factor {factor})")
        elif choice == '6':
            print("Goodbye!")
            break
        else:
            print("Invalid choice.")

if __name__ == "__main__":
    main()

File: data_converter.go
// data_converter.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type UnitInfo struct {
	Name   string
	Factor float64
}

var units = map[string]UnitInfo{
	"b":   {"bit", 1.0 / 8},
	"B":   {"byte", 1},
	"KB":  {"kilobyte", 1000},
	"KiB": {"kibibyte", 1024},
	"MB":  {"megabyte", 1000 * 1000},
	"MiB": {"mebibyte", 1024 * 1024},
	"GB":  {"gigabyte", 1000 * 1000 * 1000},
	"GiB": {"gibibyte", 1024 * 1024 * 1024},
	"TB":  {"terabyte", 1000 * 1000 * 1000 * 1000},
	"TiB": {"tebibyte", 1024 * 1024 * 1024 * 1024},
	"PB":  {"petabyte", 1000 * 1000 * 1000 * 1000 * 1000},
	"PiB": {"pebibyte", 1024 * 1024 * 1024 * 1024 * 1024},
}

type Converter struct {
	precision int
	history   []string
}

func NewConverter(prec int) *Converter {
	return &Converter{precision: prec, history: []string{}}
}

func (c *Converter) parseValueUnit(s string) (float64, string, error) {
	s = strings.TrimSpace(s)
	// split last word as unit
	parts := strings.Fields(s)
	if len(parts) == 1 {
		// only number, assume bytes
		val, err := strconv.ParseFloat(parts[0], 64)
		if err != nil {
			return 0, "", err
		}
		return val, "B", nil
	}
	// join all but last as value
	valStr := strings.Join(parts[:len(parts)-1], " ")
	unitStr := strings.ToUpper(parts[len(parts)-1])
	val, err := strconv.ParseFloat(valStr, 64)
	if err != nil {
		return 0, "", err
	}
	// normalize unit
	for u := range units {
		if strings.EqualFold(u, unitStr) {
			return val, u, nil
		}
	}
	// try alias: if unit ends with 'b' and not in map, try without 'b' etc.
	if strings.HasSuffix(unitStr, "B") && len(unitStr) > 1 {
		short := unitStr[:len(unitStr)-1]
		for u := range units {
			if strings.EqualFold(u, short) {
				return val, u, nil
			}
		}
	}
	return 0, "", fmt.Errorf("unknown unit: %s", unitStr)
}

func (c *Converter) toBytes(value float64, unit string) float64 {
	return value * units[unit].Factor
}

func (c *Converter) fromBytes(bytesVal float64, unit string) float64 {
	return bytesVal / units[unit].Factor
}

func (c *Converter) convert(value float64, fromUnit, toUnit string) float64 {
	bytesVal := c.toBytes(value, fromUnit)
	return c.fromBytes(bytesVal, toUnit)
}

func (c *Converter) autoConvert(value float64, unit string) (float64, string) {
	bytesVal := c.toBytes(value, unit)
	bestUnit := "B"
	bestVal := bytesVal
	// iterate units in order of size (smallest to largest)
	order := []string{"b", "B", "KB", "KiB", "MB", "MiB", "GB", "GiB", "TB", "TiB", "PB", "PiB"}
	for _, u := range order {
		f := units[u].Factor
		v := bytesVal / f
		if v >= 1 && v < 1000 {
			bestUnit = u
			bestVal = v
			break
		}
	}
	// if all <1, use bits
	if bestUnit == "B" && bytesVal < 1 {
		bestUnit = "b"
		bestVal = bytesVal / units["b"].Factor
	}
	return bestVal, bestUnit
}

func (c *Converter) format(val float64, unit string) string {
	return fmt.Sprintf("%.*f %s", c.precision, val, unit)
}

func (c *Converter) addHistory(value float64, fromUnit string, result float64, toUnit string) {
	c.history = append(c.history, fmt.Sprintf("%s %s -> %s %s", c.format(value, fromUnit), fromUnit, c.format(result, toUnit), toUnit))
	if len(c.history) > 20 {
		c.history = c.history[1:]
	}
}

func (c *Converter) showHistory() {
	if len(c.history) == 0 {
		fmt.Println("No conversions yet.")
		return
	}
	fmt.Println("\n--- Conversion History (last 20) ---")
	for i, h := range c.history {
		fmt.Printf("%d. %s\n", i+1, h)
	}
}

func main() {
	converter := NewConverter(2)
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("=== Data Unit Converter ===")
	for {
		fmt.Println("\n1. Convert single value")
		fmt.Println("2. Batch convert from file")
		fmt.Println("3. Show conversion history")
		fmt.Println("4. Set precision (current:", converter.precision, ")")
		fmt.Println("5. Help / unit info")
		fmt.Println("6. Exit")
		fmt.Print("Choose: ")
		scanner.Scan()
		choice := strings.TrimSpace(scanner.Text())
		switch choice {
		case "1":
			fmt.Print("Enter value and unit (e.g., 1024 MB): ")
			scanner.Scan()
			inp := scanner.Text()
			val, fromUnit, err := converter.parseValueUnit(inp)
			if err != nil {
				fmt.Println("Error:", err)
				continue
			}
			fmt.Print("Target unit (leave blank for auto): ")
			scanner.Scan()
			target := strings.TrimSpace(scanner.Text())
			if target != "" {
				toUnit := ""
				for u := range units {
					if strings.EqualFold(u, target) {
						toUnit = u
						break
					}
				}
				if toUnit == "" {
					fmt.Println("Unknown target unit.")
					continue
				}
				result := converter.convert(val, fromUnit, toUnit)
				converter.addHistory(val, fromUnit, result, toUnit)
				fmt.Printf("\nResult: %s = %s\n", converter.format(val, fromUnit), converter.format(result, toUnit))
			} else {
				result, bestUnit := converter.autoConvert(val, fromUnit)
				converter.addHistory(val, fromUnit, result, bestUnit)
				fmt.Printf("\nResult: %s = %s\n", converter.format(val, fromUnit), converter.format(result, bestUnit))
			}
		case "2":
			fmt.Print("Enter batch file path: ")
			scanner.Scan()
			fname := scanner.Text()
			file, err := os.Open(fname)
			if err != nil {
				fmt.Println("Error:", err)
				continue
			}
			defer file.Close()
			fmt.Print("Target unit for all conversions (leave blank for auto): ")
			scanner.Scan()
			target := strings.TrimSpace(scanner.Text())
			toUnit := ""
			if target != "" {
				for u := range units {
					if strings.EqualFold(u, target) {
						toUnit = u
						break
					}
				}
				if toUnit == "" {
					fmt.Println("Unknown target unit.")
					continue
				}
			}
			fileScanner := bufio.NewScanner(file)
			fmt.Println("\nBatch results:")
			for fileScanner.Scan() {
				line := strings.TrimSpace(fileScanner.Text())
				if line == "" {
					continue
				}
				val, fromUnit, err := converter.parseValueUnit(line)
				if err != nil {
					fmt.Printf("Skipping '%s': %v\n", line, err)
					continue
				}
				if toUnit != "" {
					result := converter.convert(val, fromUnit, toUnit)
					fmt.Printf("%s -> %s\n", converter.format(val, fromUnit), converter.format(result, toUnit))
				} else {
					result, bestUnit := converter.autoConvert(val, fromUnit)
					fmt.Printf("%s -> %s\n", converter.format(val, fromUnit), converter.format(result, bestUnit))
				}
			}
		case "3":
			converter.showHistory()
		case "4":
			fmt.Print("Enter number of decimal places: ")
			scanner.Scan()
			prec, err := strconv.Atoi(strings.TrimSpace(scanner.Text()))
			if err != nil || prec < 0 {
				fmt.Println("Invalid precision.")
			} else {
				converter.precision = prec
				fmt.Println("Precision updated.")
			}
		case "5":
			fmt.Println("\nSupported units:")
			for u, info := range units {
				fmt.Printf("  %4s = %s (factor %.0f)\n", u, info.Name, info.Factor)
			}
		case "6":
			fmt.Println("Goodbye!")
			return
		default:
			fmt.Println("Invalid choice.")
		}
	}
}
